tools/ci/tc/taskgraph.py

```diff
@@ -0,0 +1,170 @@
+import json
+import os
+import re
+from copy import deepcopy
+
+import six
+import yaml
+from six import iteritems
+
+here = os.path.dirname(__file__)
+
+
+def first(iterable):
+    # First item from a list or iterator
+    if not hasattr(iterable, "next"):
+        if hasattr(iterable, "__iter__"):
+            iterable = iter(iterable)
+        else:
+            raise ValueError("Object isn't iterable")
+    return next(iterable)
+
+
+def load_task_file(path):
+    with open(path) as f:
+        return yaml.safe_load(f)
+
+
+def update_recursive(data, update_data):
+    for key, value in iteritems(update_data):
+        if key not in data:
+            data[key] = value
+        else:
+            initial_value = data[key]
+            if isinstance(value, dict):
+                if not isinstance(initial_value, dict):
+                    raise ValueError("Variable %s has inconsistent types "
+                                     "(expected object)" % key)
+                update_recursive(initial_value, value)
+            elif isinstance(value, list):
+                if not isinstance(initial_value, list):
+                    raise ValueError("Variable %s has inconsistent types "
+                                     "(expected list)" % key)
+                initial_value.extend(value)
+            else:
+                data[key] = value
+
+
+def resolve_use(task_data, templates):
+    rv = {}
+    if "use" in task_data:
+        for template_name in task_data["use"]:
+            update_recursive(rv, deepcopy(templates[template_name]))
+    update_recursive(rv, task_data)
+    rv.pop("use", None)
+    return rv
+
+
+def resolve_name(task_data, default_name):
+    if "name" not in task_data:
+        task_data["name"] = default_name
+    return task_data
```

Comment:
I found it a bit surprising that a method named `resolve_name` doesn't return the name, but next to `resolve_use` it makes sense for consistency.